File: src/main/java/com/vc/hard/L1136.scala
package com.vc.hard

object L1136 {
  def minimumSemesters(N: Int, arr: Array[Array[Int]]): Int = {
    import scala.collection.mutable
    val map = new mutable.HashMap[Int, mutable.HashSet[Int]]()
    val indegree = new Array[Int](N + 1)
    arr.indices.foreach(i => {
      val relation = arr(i)
      val set = map.getOrElse(relation(0), new mutable.HashSet[Int]())
      set += relation(1)
      map.put(relation(0), set)
      indegree(relation(1)) += 1
    })

    val q = new mutable.Queue[Int]()
    (1 to N).foreach(i => {
      if(indegree(i) == 0) q.enqueue(i)
    })

    var res = 0
    val visited = new mutable.HashSet[Int]()
    while(q.nonEmpty) {
      var s = q.size
      while(s > 0) {
        val e = q.dequeue()
        if(!visited.contains(e)) {
          visited += e
          if(map.contains(e)) {
            map(e).foreach(listElement => {
              indegree(listElement) -= 1
              if(indegree(listElement) == 0) q.enqueue(listElement)
            })
          }
        }
        s -= 1
      }
      res += 1
    }
    if(indegree.count(_ == 0) == N + 1) res else -1
  }
}

File: src/main/java/com/vc/hard/L1136.java
package com.vc.hard;

import java.util.*;

class L1136 {
    public int minimumSemesters(int N, int[][] relations) {
        HashMap<Integer, Integer> indegree = new HashMap<>();
        HashMap<Integer, List<Integer>> map = new HashMap<>();

        for(int i = 1; i <= N; i++) indegree.put(i, 0);

        for(int[] relation: relations) {
            int from = relation[0];
            int to = relation[1];

            indegree.put(to, indegree.get(to) + 1);
            List<Integer> list = map.getOrDefault(from, new ArrayList<>());
            list.add(to);
            map.put(from, list);
        }

        //System.out.println(indegree);
        //System.out.println(map);

        int semester = 0;
        Queue<Integer> q = new LinkedList<>();
        for(Map.Entry<Integer, Integer> entry: indegree.entrySet()) {
            if(entry.getValue() == 0) {
                q.offer(entry.getKey());
                N--;
            }
        }

        while(!q.isEmpty()) {
            int size = q.size();
            for(int i = 0; i < size; i++) {
                int course = q.poll();
                if(!map.containsKey(course)) continue;
                for(int courseAfter: map.get(course)) {
                    indegree.put(courseAfter, indegree.getOrDefault(courseAfter, 0) - 1);
                    if(indegree.get(courseAfter) == 0) {
                        q.offer(courseAfter);
                        N--;
                    }
                }
            }
            semester++;
        }

        return N == 0 ? semester : -1;
    }
}
